dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-c3/src/main/java/org/dashbuilder/renderer/c3/client/charts/map/widgets/D3Map.java

```diff
@@ -0,0 +1,260 @@
+    
+    private int width;
+    private int height;
+    private D3MapConf conf;
+    private D3.Scale colorScale;
+    private GeoData geoData;
+    private Map<String, Double> data;
+    
+    @Inject
+    private MapTooltip mapTooltip;
+    
+    @Inject
+    private HTMLDivElement mapContainer;
+
+    D3 d3 = D3.Builder.get();
+    String[] COLORS_SCHEME = d3.getSchemeGreens()[7];
+
+    @PostConstruct
+    public void init() {
+        addTooltipElement();
+    }
+    
+    @Override
+    public HTMLElement getElement() {
+        return mapContainer;
+    }
+    
+    public void generateMap(int width, int height, D3MapConf conf) {
+        mapContainer.innerHTML = "";
+        this.width = width;
+        this.height = height;
+        this.conf = conf;
+        this.geoData = conf.getGeoData();
+        this.data = conf.getData();
+        Element mapSVG = DomGlobal.document.createElementNS("http://www.w3.org/2000/svg", "svg");
+        Selection d3Selection = d3.select(mapSVG);
+        D3PathGenerator pathGenerator = createPathGenerator();
+        
+        IntSummaryStatistics statistics = data.values().stream()
+                                                                 .mapToInt(v -> v.intValue()).summaryStatistics();
+        Integer[] domain = new Integer[] {statistics.getMin(), statistics.getMax()};
+        colorScale = d3.scaleQuantize()
+                                    .domain(domain)
+                                    .range(COLORS_SCHEME);
+        
+        D3 map = createMap(d3Selection, pathGenerator);
+        
+        if (conf.isRegions()) {
+            fillRegions(map);
+        }
+        if (conf.isMarkers()) {
+            addMarkers(d3Selection, pathGenerator);
+        }
+        createLegend(d3Selection);
+        mapContainer.appendChild(mapSVG);
+    }
+    
+    private D3 createMap(Selection d3Selection, D3PathGenerator pathGenerator) {
+        JSONArray countriesFeatures = geoData.countries();
+        return d3Selection.attr("width", width)
+                                      .attr("height", height)
+                                      .style("background", conf.getBackgroundColor())
+                                      .append("g")
+                                      .attr("class", "countries")
+                                      .selectAll("path")
+                                      .data(countriesFeatures.getJavaScriptObject())
+                                      .enter().append("path")
+                                      .attr("d", pathGenerator);
+    }
+
+    private void createLegend(D3 d3Selection) {
+        int titleSize = 5;
+        int legendSquareSize = 12;
+        int legendSize = titleSize + (legendSquareSize * COLORS_SCHEME.length);
+        AtomicInteger rectPos = new AtomicInteger(); 
+        AtomicInteger textPos = new AtomicInteger(legendSquareSize); 
+        String translate = "translate(0, " + (height - legendSize - 2) + ")";
```

Comment:
Maybe it makes sense to protect against this becoming negative?
`Math.max(0 , height - legendSize - 2);`

Reply:
Yes, it will be nevative only if users define really small maps. I did it "responsible" to the map size, but the legends where getting weird with larger maps, so I decided to keep it fixed. However, if it is negative it means that the legend would be overridden the map completely, hence the legend would be moved outside the visible area, hence I think it is harmful